### PR TITLE
BGDIINF_SB-3129: Replaced short_id generator #major

### PR DESCRIPTION
**Pull request by ltflb-bgdi**
Replaced timestamp based id generator by nanoid
(https://github.com/puyuan/py-nanoid)

Added config parameters for id-size and alphabet.
Refer to README for more info.

Upgraded Python to 3.12

Made linter happy (circular import issue yet to be fixed)

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -4,21 +4,22 @@ verify_ssl = true
...
-logging-utilities = "~=3.0"
-Flask = "~=2.1.0"
-gevent = "~=21.12.0"
-gunicorn = "~=20.1.0"
-PyYAML = ">=5.4"
-python-dotenv = "~=0.20.0"
-validators = "~=0.19.0"
+boto3 = "~=1.28"
+logging-utilities = "~=4.0"
+Flask = "~=3.0"
```
Comment (ltshb) [inline Pipfile:9]: Did you check if they are any breaking change in version 3 ?

File: app/settings.py [hunk 2]
```diff
@@ -26,6 +26,8 @@
 STAGING = os.environ['STAGING']
 
 COLLISION_MAX_RETRY = 10
-SHORT_ID_SIZE = 10
+
+SHORT_ID_SIZE = os.getenv('SHORT_ID_SIZE', '12')
```
Comment (ltshb) [inline app/settings.py:30]: I would do the conversion to `int` and not in the code.

File: tests/unit_tests/test_routes.py [hunk 3]
```diff
@@ -35,9 +35,16 @@ def test_create_shortlink_ok(self):
         shorturl = response.json.get('shorturl')
         self.assertEqual('http://localhost/' in shorturl, True)
         short_id = shorturl.replace('http://localhost/', '')
-        self.assertIsNotNone(
-            re.search("^[0-9A-Za-z-_]{" + str(SHORT_ID_SIZE) + "}$", short_id),
-            msg=f'Short ID {short_id} doesn\'t match regex'
+        self.assertEqual(
+            len(short_id),
+            int(SHORT_ID_SIZE),
```
Comment (ltshb) [inline tests/unit_tests/test_routes.py:40]: See comment above

File: tests/unit_tests/test_routes.py [hunk 4]
```diff
@@ -35,9 +35,16 @@ def test_create_shortlink_ok(self):
...
+            msg=
+            f"Length of short_id '{short_id}' does not match configured size of {SHORT_ID_SIZE} characters"  # pylint: disable=line-too-long
```
Comment (ltshb) [inline tests/unit_tests/test_routes.py:42]: Breaks the line in 2 here and don't disable pylint warning. IMHO the only reason to disable line-too-long pylint would be for regex and urls.

```python
msg=f"Length of short_id '{short_id}' does not match configured size " 
f"of {SHORT_ID_SIZE} characters"
```

File: app/helpers/utils.py [hunk 5]
```diff
@@ -70,8 +71,7 @@ def get_redirect_param(ignore_errors=False):
 
 
 def generate_short_id():
-    # datetime.datetime(2001, 9, 9, 3, 46, 40) * 1000 = 1000000000000
-    return f'{int(time.time() * 1000) - 1000000000000:x}'
+    return generate(SHORT_ID_ALPHABET, int(SHORT_ID_SIZE))
```
Comment (ltshb) [inline app/helpers/utils.py:74]: See comment below

**Review by ltshb — APPROVED** [review]
Personally because 3.12 has just been released, I would have updated to 3.11 instead.

File: app/routes.py [hunk 6]
```diff
@@ -51,16 +51,16 @@ def create_shortlink():
...
     the information about the shortlink.
     """
-    should_redirect = get_redirect_param()
+
     db_entry = get_db().get_entry_by_shortlink(shortlink_id)
     if db_entry is None:
         abort(404, f'No short url found for {shortlink_id}')
 
-    if should_redirect:
+    if get_redirect_param():
```
Comment (hansmannj) [inline app/routes.py:63]: Nice :wink:

File: tests/unit_tests/test_routes.py (hunk 4 above)
Comment (hansmannj) [reply tests/unit_tests/test_routes.py:42]: :+1:
When I saw the pylint pragma here, I could almost hear @ltshb voice mentioning, to better split the line than disabling the warning :wink:

**Review by hansmannj — APPROVED** [review]
Nice, thanks!
And yes, as @ltshb says, probably we'd stay with 3.11 for a while, before switching to 3.12

File: app/settings.py (hunk 2 above)
Comment (ltflb-bgdi) [reply app/settings.py:30]: Thanks. Already did in the meantime

File: Pipfile (hunk 1 above)
Comment (ltflb-bgdi) [reply Pipfile:9]: As far as I can see there are no breaking changes affecting the service

File: tests/unit_tests/test_routes.py (hunk 3 above)
Comment (ltflb-bgdi) [reply tests/unit_tests/test_routes.py:40]: Done

File: tests/unit_tests/test_routes.py (hunk 4 above)
Comment (ltflb-bgdi) [reply tests/unit_tests/test_routes.py:42]: Done

File: app/helpers/utils.py (hunk 5 above)
Comment (ltflb-bgdi) [reply app/helpers/utils.py:74]: Done

**Review by ltshb — APPROVED** [review]
👍🏼

**Pull request merged**